libdill.h

```diff
@@ -967,6 +967,75 @@ DILL_EXPORT int dill_ws_response_key(
 #define ws_response_key dill_ws_response_key
 #endif
 
+/******************************************************************************/
+/*  SOCKS5                                                                    */
+/******************************************************************************/
+
+// SOCKS5 client commands
+#define DILL_SOCKS5_CONNECT (0x01)
+#define DILL_SOCKS5_BIND (0x02)
+#define DILL_SOCKS5_UDP_ASSOCIATE (0x03)
+
+// SOCKS5 server reply codes
+#define DILL_SOCKS5_SUCCESS (0x00)
+#define DILL_SOCKS5_GENERAL_FAILURE (0x01)
+#define DILL_SOCKS5_CONNECTION_NOT_ALLOWED (0x02)
+#define DILL_SOCKS5_NETWORK_UNREACHABLE (0x03)
+#define DILL_SOCKS5_HOST_UNREACHABLE (0x04)
+#define DILL_SOCKS5_CONNECTION_REFUSED (0x05)
+#define DILL_SOCKS5_TTL_EXPIRED (0x06)
+#define DILL_SOCKS5_COMMAND_NOT_SUPPORTED (0x07)
+#define DILL_SOCKS5_ADDRESS_TYPE_NOT_SUPPORTED (0x08)
+
+
+typedef int dill_socks5_auth_function(const char *username,
+    const char *password);
+
+DILL_EXPORT int dill_socks5_client_connect(
+    int s, const char *username, const char *password,
+    struct dill_ipaddr *addr, int64_t deadline);
+
+DILL_EXPORT int dill_socks5_client_connectbyname(
+    int s, const char *username, const char *password, const char *hostname,
+    int port, int64_t deadline);
+
+DILL_EXPORT int dill_socks5_proxy_auth(
+    int s, dill_socks5_auth_function *auth_fn, int64_t deadline);
```

Comment:
The callback is ugly. Can we just have two functions like "recvrequest" which would return all data sent by the client, including name & password, and then "sendresponse" ?

Reply:
Yeah, I'm not a big fan of callbacks either, but the user/password authentication needs to be flexible as the values could be static, or coming from a database, or validated by an external API. 

Having an intermediate return point mid-auth would mean you're stopping mid-protocol which would then require passing context as the responses are slightly different (i.e. there are two intermediate states which are literally from different RFCs). Specifically, User/Password authentication is optional. In the case of NO AUTH the response based on the RFC 1928 version (version code 5). In the case of USER/PASS auth, the response is from RFC1929 (which has version code 1). As it happens, the success (0x00) and failure (0xFF) codes match. (and then you have more API calls, not less). 

Rolling it all together, the client won't send the request until after it has completed authentication. I suppose you could always accept whatever authentication the client provides and then reply back to the request with SOCKS5_CONNECTION_NOT_ALLOWED. Then end result would be the same (an error and closing the connection), but then the client is getting the wrong error condition. That seems more wrong as it's likely to result in users getting misleading error responses (instead of something helpful like "dude, your username/password is wrong").

I'm not sure there is a "right" answer here (besides a time-machine-rewrite-RFCs answer). To me, the callback is the "less wrong" answer. 

Combining auth and recvcommand (still including the callback) is certainly possible. I thought about this. Essentially you'd just be adding one parameter to recvcommand (the auth callback function). something like:

err = socks5_proxy_auth_recvcommandbyname(s, auth_fn, &addr, deadline)

I think its important to convey that it is performing auth + recvcommand. The name is starting to sound like a run-on sentence. Thoughts?

Reply:
I think the real question is where to draw the line between "protocol implementation" (should be in libdill) and "application implementation" (should be in user's code). That in turn boils down to the question "what exactly could be reused among different kinds of socks5 servers?"

One way to answer that question is how http implementation did it, i.e. it just concerns itself with sending and receiving pieces of the protocol on the wire, leaving everything else (including correct sequencing of the pieces) to the application. That was what I was trying to propose. 

Another way to think about it is that if the protocol has a single handshake, there should be only 2 functions on the server side: recv request, send response.

But the details matter, so I'm going to read the RFC once I have some free time.